src/app/utils/generateEmptyStateMsg.tsx

```diff
@@ -0,0 +1,23 @@
+import type { ReactNode } from "react";
+
+import { Spinner } from "@canonical/react-components";
+
+export const generateEmptyStateMsg = ({
+  emptyStateMsg,
+  isLoading,
+  hasFilter,
+  emptySearchMsg = "No data matches the search criteria",
+}: {
+  emptyStateMsg: ReactNode;
+  isLoading?: boolean;
+  hasFilter?: boolean;
+  emptySearchMsg?: ReactNode;
+}): ReactNode => {
+  if (isLoading) {
+    return <Spinner text="Loading..." />;
+  } else if (hasFilter) {
+    return emptySearchMsg;
+  }
+
+  return emptyStateMsg;
+};
```

Comment:
Nice work on this util. I have a few suggestions for improvements.

It seems like there are two intertwined concepts here: generating labels and determining status based on booleans. This can make it code hard to understand and maintain. 

A potential solution could be to create a separate utility to obtain a single status string from the booleans and use an object to map the status to its corresponding label. This would clearly separate the two.

```tsx
const getTableStatus = ({
  isLoading,
  hasFilter,
}: {
  isLoading?: boolean;
  hasFilter?: boolean;
}): "loading" | "filtered" | "default" => {
  if (isLoading) {
    return "loading";
  }
  if (hasFilter) {
    return "filtered";
  }
  return "default";
};

export const generateEmptyStateMsg = (
  status: "loading" | "filtered" | "default",
  customLabelMapping: Partial<{
    loading: ReactNode;
    filtered: ReactNode;
    default: ReactNode;
  }> = {}
): ReactNode => {
  const defaultLabelMapping = {
    loading: <Spinner text="Loading..." />,
    filtered: "No data matches the search criteria",
    default: "No data is available",
  };

  const labelMapping = {
    ...defaultLabelMapping,
    ...customLabelMapping,
  };

  return labelMapping[status];
};
```

It would allow for simpler usage, like this:

```ts
const statusString = getTableStatus({ isLoading, hasFilter });

const customMessages = {
  filtered: "Custom search criteria message",
};

// Usage with default labels
const defaultMessage = generateEmptyStateMsg(statusString);

// Usage with partial custom labels
const customMessage = generateEmptyStateMsg(statusString, customMessages);
```

Reply:
We should also make it clear that this util is meant to be used only in tables (e.g. by placing it in a table-related folder/file).

Reply:
This looks a lot cleaner, thanks for your insight